src/main/java/com/hubspot/ringleader/watcher/WatcherFactory.java

```diff
@@ -1,45 +1,234 @@
+      return;
+    }
+
+    executor.submit(new Runnable() {
 
       //@Override Java 5 compatibility
-      public void newEvent(Event event) {
-        started.countDown();
+      public void run() {
+        CuratorFramework previous = curatorReference.getAndSet(createNewCurator());
+        cleanup(previous);
+        if (runnable != null) {
+          runnable.run();
+        }
       }
     });
+  }
 
-    return watcher;
+  private void startIfNecessary() {
+    if (started.compareAndSet(false, true)) {
+      curatorReference.set(createNewCurator());
+    }
+  }
+
+  private CuratorFramework createNewCurator() {
+    CuratorFramework curator = null;
+    try {
+      curator = curatorSupplier.get();
+      if (curator.getState() != CuratorFrameworkState.STARTED) {
+        curator.start();
+      }
+
+      curator.getConnectionStateListenable().addListener(new ConnectionStateListener() {
+
+        //@Override Java 5 compatibility
+        public void stateChanged(CuratorFramework client, ConnectionState newState) {
+          switch (newState) {
+            case SUSPENDED:
+            case LOST:
+              LOG.error("Connection lost or suspended, replacing client");
+              replaceCurator(null);
+              break;
+            default:
+              // make findbugs happy
+          }
+        }
+      });
+
+      curator.getUnhandledErrorListenable().addListener(new UnhandledErrorListener() {
+
+        //@Override Java 5 compatibility
+        public void unhandledError(String message, Throwable e) {
+          LOG.error("Curator error, replacing client", e);
+          replaceCurator(null);
+        }
+      });
+
+      return curator;
+    } catch (Exception e) {
+      LOG.error("Error creating curator", e);
+      cleanup(curator);
+      return null;
+    }
+  }
+
+  private void cleanup(final CuratorFramework curator) {
+    if (curator == null) {
+      return;
+    }
+
+    executor.submit(new Runnable() {
+      //@Override Java 5 compatibility
+      public void run() {
+        synchronousCleanup(curator);
+      }
+    });
+  }
+
+  private void synchronousCleanup(final CuratorFramework curator) {
+    if (curator == null) {
+      return;
+    }
+
+    try {
+      try {
+        curator.close();
+      } catch (UnsupportedOperationException e) {
+        // NamespaceFacade throws UnsupportedOperationException when you try to close it
+        // Need to resort to reflection to access real CuratorFramework instance so we can close it
+        Field curatorField = curator.getClass().getDeclaredField("client");
+        curatorField.setAccessible(true);
+
+        synchronousCleanup((CuratorFramework) curatorField.get(curator));
+      }
+    } catch (Exception e) {
+      LOG.debug("Error closing curator", e);
+    }
+  }
+
+  private static ScheduledExecutorService getExecutor() {
```

Comment:
maybe call this `newExecutor` to make it clear it creates an executor vs. just returning an existing one